test/object-store/sync/flx_role_change.cpp

```diff
@@ -146,20 +147,33 @@ void verify_records(SharedRealm& check_realm, size_t emps, size_t mgrs, size_t d
     auto role_col = table->get_column_key("role");
     auto table_query = Query(table).equal(role_col, "employee");
     auto results = Results(check_realm, table_query);
-    CHECK(results.size() == emps);
+    REQUIRE(results.size() == emps);
     table_query = Query(table).equal(role_col, "manager");
     results = Results(check_realm, table_query);
-    CHECK(results.size() == mgrs);
+    REQUIRE(results.size() == mgrs);
     table_query = Query(table).equal(role_col, "director");
     results = Results(check_realm, table_query);
-    CHECK(results.size() == dirs);
+    REQUIRE(results.size() == dirs);
+}
+
+// Helper lambda to wait for realm download/upload/advance and then validate the record
+// counts in the local realm.
+void wait_and_verify(SharedRealm realm, size_t emps, size_t mgrs, size_t dirs)
+{
+    // Using a bool to check the wait results, since REQUIRE was causing TSAN errors
+    // with the REQUIRE calls in the event hook
+    bool success = !wait_for_download(*realm);
+    success = success && !wait_for_upload(*realm);
+    if (!success)
+        FAIL("Failed to update realm");
```

Comment:
This will produce the same data races that REQUIRE would. Catch2 isn't internally thread-safe, so we have to ensure that we're only using it from one thread at a time.

Reply:
Now it is returning the `wait_for_download()` or `wait_for_upload()` result if it fails and the bool is checked under a mutex where needed.